shaptools/hdb_connector/connectors/dbapi_connector.py

```diff
@@ -65,11 +65,18 @@ def query(self, sql_statement):
         self._logger.info('executing sql query: %s', sql_statement)
         try:
             with self._connection.cursor() as cursor:
-                cursor.execute(sql_statement)
-                result = base_connector.QueryResult.load_cursor(cursor)
+                try:
+                    cursor.execute(sql_statement)
+                except dbapi.Error as err:
```

Comment:
@arbulu89 After looking at the big picture again I would like to come up with a much simpler approach ...

- The error is raised in `dbapi_connector.py`, and originates from `hdbcli.dbapi` (where we cannot fix/change the code).
- I know of no way to fix this in `salt-shaptools` as we do not use `hdbcli` in there but only `base_connector`.
- Another place would be `base_connector.py`, which has a **"TODO: catch any exceptions raised by fetchall()"** but importing `hdbcli` there is also not a good idea.
- What about simply excepting this particular error `No result set` and declaring it an "non-error", leaving the rest untouched?

```
    def query(self, sql_statement):
        """
        Query a sql query result and return a result object
        """
        self._logger.info('executing sql query: %s', sql_statement)
        try:
            with self._connection.cursor() as cursor:
                cursor.execute(sql_statement)
                result = base_connector.QueryResult.load_cursor(cursor)
                return result
```

By simply adding this ... ?

```
        except dbapi.ProgrammingError as err:
            if err.errorcode == 0 and err.errortext == 'No result set':
                pass # This is not error, but an empty query result.
            else:
                raise err
```
```
        except dbapi.Error as err:
            raise base_connector.QueryError('query failed: {}'.format(err))
```

No other case/exception should be impacted by this and only this particular case is covered.

Reply:
For me that's not a good practice. You want to mask an actual error with something else. Not having results should mean that, not having a result, and if the driver treats this as an exception, why shouldn't we?

Why don't you capture this `dbapi.ProgrammingError` and re-raise other custom error (`QueryError` for instance)?
You can capture this later on in `salt-shaptools`.

Anyway, in the first message logs, I already see that `base_connector.QueryError` is raised. Cannot we just capture it?

What you can do is in `salt-shaptools`, if you have a `QueryError` and the message is `No result set`, you don't raise the error. But this happens on the consumer side. Otherwise, we would lose flexibility. Something like:

```
except base_connector.QueryError as e:
    if str(e) != "query failed: (0, 'No result set')":
        raise exceptions.Command....
```